src/interpreter.ts

```diff
@@ -616,65 +74,8 @@ export function parseAndEvalExpression(sourceCode: string): EvalResult {
     }
 }
 
-const defaultInterpreterConfig: InterpreterConfig = {
-    // We set the default max number of loop iterations
-    // to the maximum number allowed for repeat loops
-    maxLoopIterations: maxRepeatStatement,
-};
-
-/*
-Interprets Tact AST trees. 
-The constructor receives an optional CompilerContext which includes 
-all external declarations that the interpreter will use during interpretation.
-If no CompilerContext is provided, the interpreter will use an empty 
-CompilerContext.
-
-**IMPORTANT**: if a custom CompilerContext is provided, it should be the 
-CompilerContext provided by the typechecker. 
-
-The reason for requiring a CompilerContext is that the interpreter should work 
-in the use case where the interpreter only knows part of the code.
-For example, consider the following code (I marked with brackets [ ] the places 
-where the interpreter gets called during expression simplification in the 
-compilation phase):
-
-const C: Int = [1];
-
-contract TestContract {
-
-   get fun test(): Int {
-      return [C + 1];
-   }
-}
-
-When the interpreter gets called inside the brackets, it does not know what 
-other code is surrounding those brackets, because the interpreter did not execute the 
-code outside the brackets. Hence, it relies on the typechecker to receive the 
-CompilerContext that includes the declarations in the code 
-(the constant C for example).
-
-Since the interpreter relies on the typechecker, it assumes that the given AST tree
-is already a valid Tact program.
-
-Internally, the interpreter uses a stack of environments to keep track of
-variables at different scopes. Each environment in the stack contains a map
-that binds a variable name to its corresponding value.
-*/
-export class Interpreter {
-    private envStack: EnvironmentStack;
-    private context: CompilerContext;
-    private config: InterpreterConfig;
-
-    constructor(
-        context: CompilerContext = new CompilerContext(),
-        config: InterpreterConfig = defaultInterpreterConfig,
-    ) {
-        this.envStack = new EnvironmentStack();
-        this.context = context;
-        this.config = config;
-    }
-
-    public interpretModuleItem(ast: AstModuleItem): void {
+export abstract class InterpreterInterface<T> {
```

Comment:
OOP in Typescript, and `abstract` in particular are _heavily_ unsound. If that wasn't bad enough, they're broken even in Javascript, where `o.f()` and `(o).f()` have different meaning, and to add insult to injury this kind of problems are not checked by Typescript at all.

Abstract interpreter is best expressed using "tagless final" pattern. Here's a simple example.

Imagine we have a language of numbers and addition. We define

```ts
interface Expr<T> {
    num: (value: number) => T;
    add: (left: T, right: T) => T;
}
```

and can now express `1 + (2 + 3)` as

```ts
const example = <T>({ num, add }: Expr<T>) => add(num(1), add(num(2), num(3)));
```

Then we can define abstract interpreters for `Expr`:

```ts
const show: Expr<string> = {
    num: (value) => `${value}`,
    add: (left, right) => `${left} + ${right}`,
};

const run: Expr<number> = {
    num: (value) => value,
    add: (left, right) => left + right,
};
```

The only weird thing here is that whenever we apply an interpreter to a value, it's actually applying a value to interpreter: `example(show)`. If that makes anyone uncomfortable, it's easy to have another wrapper

```ts
const showAlg: Expr<string> = { ... };
const show = (value: (algebra: Expr<string>) => string): string => value(showAlg);
```

From a theory side of things, this is just a regular encoding for existential type through double negation of universal quantification.

Now to more obscure parts of this encoding. You could have mentioned that `show` would display `1 + 2 + 3` instead of `1 + (2 + 3)`, and we'll make a scarecrow issue out of it. Let's add some context that keeps a boolean flag of whether subexpression should be wrapped into parens.

```ts
const showAlg: Expr<(topLevel: boolean) => string> = {
    num: (value) => () => `${value}`,
    add: (left, right) => (topLevel) => {
        const tmp = `${left(false)} ${right(false)}`;
        return topLevel ? tmp : `(${tmp})`;
    },
};
const show = (value) => value(showAlg)(true);
```

That kind of Reader monad pattern happens most of the times we intend to use tagless final encoding, and in fact it's very often that we want to add a few other values into our context. In order not to face the tedium of adjusting every single place of our program, it's best to make it an object the first time the code is implemented

```ts
type FooContext = { flag: boolean }; // only one... for now
type Foo = (ctx: FooContext) => string;
const fooAlg: Expr<Foo> = { ... };
```

The best part of the pattern (and in fact the reason it was initially invented) comes from modularity of these interpreters. We have both sequential and parallel composition here:

```
const showButAll1: Expr<string> = {
   ...show,
   num: () => '1',
};

const withLogging = <T>({ num, add }: Expr<T>): Expr<(log: () => string) => T> => ({
    num: (value) => (log) => { log(value.tostring()); return num(value); },
    add: (left, right) => (log) => add(left(log), right(log)),
});
const showWithLog = withLogging(show);

const translate = <T>(algebra: Expr1<T>): Expr2<T> => ({ ... });
```

and also we can extend (or even reduce) the language without modifying any and all of the previously working code. Together these two solve the [extension problem](https://inria.hal.science/hal-02399166/file/The_Software_Language_Extension_Problem.pdf)

```ts
interface ExtendedExpr<T> extend Expr<T> {
    mul: (left: T, right: T) => T;
}

const showExtended: ExtendedExpr<string> = {
    ...show,
    mul: (left, right) => `${left} * ${right}`,
};

// none of original code was changed
```

When this pattern does not work that well:
- If we need any operations with more than 1 argument (for example, equality), it takes huge effort to encode it, and even then we're guaranteed (proofs omitted) to spend at least `O(N^2)` operations. I wish Kiselyov mentioned it in original paper instead of a passing mention in related source code in his blog :/
- If we have to consider more than one node at a time. Usually it signifies we have our ast types wrong, but we already have codebase with whatever ast is there, and I suspect you're not a fan of the idea of refactoring it. It's definitely possible to use context to keep the information of nodes around the current. Like, when we want to do something on node of type B inside of node of type A, we just add `insideA: boolean` flag. It's just a bit tedious, so general rule is not to use tagless final whenever lots of pattern-matching is expected. I should probably mention there is a Boehm-Berarducci encoding for arbitrary pattern-matching, but it doesn't look very practical to me.

Reply:
Now that I described in some detail what might be done here, let's come back to the topic of OOP and where it makes sense to use it (I mean, anything that has `class` or `new` in it).

The short answer would be "never", but it's not entirely true. In `prototype`s js holds only one object with all the method pointers, and when we _do not_ use a class, we have to create a closure object for every method in every instance of the "class". In fact it's an exceptionally rare situation when an object holds more than a couple of methods (very likely they should have been just standalone functions), or where extra memory usage and runtime overhead are not worth extra soundness. It usually happens in a code that operates with billions of objects.

Reply:
> The short answer would be "never", but it's not entirely true

Sorry if I'll interrupt your thoughts, but I want to add that using a minimal amount of OOP is very handy when dealing with the Object Pool pattern where you minimize new memory allocations and greatly reduce GC spikes by re-using already created objects. And for those to have a method like `.reset()` that'll operate on the fields is super handy :)